### ISSUE TITLE
Unable to set cursor

### ISSUE THREAD

**Issue by wendao8469759** [spring-projects/spring-data-redis#2584]
I have searched all the information and found that scan cannot set cursor and setting count is invalid. What is the correct way to write it here?

**Comment by wendao8469759** [comment]
```
 private Set<String> scanTargetKeys(String keyWords, Integer pageNum,Integer pageSize) {
        if (StringUtil.isEmpty(keyWords)) {
            keyWords = "*";
        }
        String finalKeyWords = keyWords;
        return redisTemplate.execute((RedisCallback<Set<String>>) connection -> {

            Set<String> keysTmp = new HashSet<>();

            String scanPointerKey = SCAN_POINTER_PREFIX + pageNum;
            byte[] scanPointerBytes = connection.get(scanPointerKey.getBytes());

            if (scanPointerBytes == null) {
                scanPointerBytes = "0".getBytes();
            }
            ScanOptions.ScanOptionsBuilder scanOptionsBuilder = ScanOptions.scanOptions();
            scanOptionsBuilder.match(finalKeyWords);
            scanOptionsBuilder.count(pageSize);

            Cursor<byte[]> cursor = connection.scan(scanOptionsBuilder.build());
            int i = 0;
            while (cursor.hasNext()) {
                String key = new String(cursor.next());
                if (i >= (pageNum - 1) * pageSize && i < pageNum * pageSize) {
                    keysTmp.add(key);
                }
                if (i >= pageNum * pageSize) {
                    break;
                }
                i++;
            }

            return keysTmp;
        });
    }
```

**Comment by wendao8469759** [comment]
```
private Set<String> scanTargetKeys(String keyWords, Integer pageNum,Integer pageSize) {
        if (StringUtil.isEmpty(keyWords)) {
            keyWords = "*";
        }
        String finalKeyWords = keyWords;
        return redisTemplate.execute((RedisCallback<Set<String>>) connection -> {
            Set<String> keysTmp = new HashSet<>();
            ScanOptions.ScanOptionsBuilder scanOptionsBuilder = ScanOptions.scanOptions();
            scanOptionsBuilder.match(finalKeyWords);
            scanOptionsBuilder.count(pageSize);

            Cursor<byte[]> cursor = connection.scan(scanOptionsBuilder.build());
            int i = 0;
            while (cursor.hasNext()) {
                String key = new String(cursor.next());
                if (i >= (pageNum - 1) * pageSize && i < pageNum * pageSize) {
                    keysTmp.add(key);
                }
                if (i >= pageNum * pageSize) {
                    break;
                }
                i++;
            }

            return keysTmp;
        });
    }
```

**Comment by jxblum** [comment]
You might want to post this "question" on StackOverflow, [here](https://stackoverflow.com/questions/tagged/spring-data-redis) to the [Spring Data] Redis community at large.

GitHub Issues for Spring Data Redis are for bugs, improvements and new features in SD Redis itself, or changes to the documentation, whether in the [reference doc](https://docs.spring.io/spring-data/redis/docs/current/reference/html) or [Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/). This is not a QA or support forum.

Having said this, I am curious why you chose to perform a `SCAN` inside a `RedisCallback` using the `RedisTemplate` rather than using the `RedisTemplate.scan(:ScanOptions)` method ([Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/core/RedisTemplate.html#scan(org.springframework.data.redis.core.ScanOptions))), specifically?

The primary difference in approach here is that with `RedisTemplate.scan(:ScanOptions)` method (unlike your (manually executed `SCAN` above) will configure and use a "sticky" connection, which is likely necessary for the application to remember the prior `Cursor` location on subsequent `CURSOR` command invocations.

If you simply invoke `RedisTemplate.execute(:RedisCallback<?>)`, then it is possible that a new connection will be acquired. Even if it is an existing (likely "pooled") connection, it is not necessarily the same connection used to perform the initial `SCAN` and therefore no prior state of the `Cursors` location would have been retained on the subsequent call.

I would simply change your code to do the following:

```java
private Set<String> scanTargetKeys(String keyWords, Integer pageNum,Integer pageSize) {

    // ...

    ScanOptions scanOptions = ScanOptions.scanOptions()
        .match(finalKeyWords)
        .count(pageSize)
        .build()

    Cursor<byte[]> cursor = redisTemplate.scan(scanOptions);

    // iterator over the Cursor as before

}    
```

Then when your `scanTargetKeys(..)` method is invoked again, this should give you the result your looking for???

If not, please provide a reproducer of the issue you are experiencing (preferably) as minimal test case or using a simple example.

**Comment by wendao8469759** [comment]
There may be some issues with my description, but I actually want to implement the Redis command: scan cursor [MATCH pattern] [COUNT count]
use cursor. next() is called, it starts iterating from the position of the previous cursor, rather than starting from 0. Currently, I use
``
Cursor<String> cursor = redisTemplate.scan(scanOptions);
cursor.next();
``
I found that every iteration starts from 0. If I have many redis keys, does it mean that I need to iterate from scratch to obtain the subsequent keys?

**Comment by wendao8469759** [comment]
```java
    private Set<String> scanTargetKeys2(String keyWords, Long pageNum, Long pageSize) {
        Set<String> keysTmp = new HashSet<>();
        if (StringUtil.isEmpty(keyWords)) {
            keyWords = "*";
        }
        String finalKeyWords = keyWords;

        ScanOptions scanOptions = ScanOptions.scanOptions()
                .match(finalKeyWords)
                .count(pageSize)
                .build();
        Cursor<String> cursor = redisTemplate.scan(scanOptions);
        while (cursor.hasNext()) {
            long cursorId = cursor.getCursorId();
            long position = cursor.getPosition();
            String key = cursor.next();
            System.out.println("cursorId = " + cursorId + " position = " + position);
            if (position >= (pageNum - 1) * pageSize && position < pageNum * pageSize) {
                keysTmp.add(key);
                if(keysTmp.size() == pageSize){
                    break;
                }
            }
            if (position >= pageNum * pageSize) {
                break;
            }
        }
        return keysTmp;
    }
```

When my pageSize=1 and pageNum=10; My loop has been executed 10 times.
When my pageSize=2 and pageNum=10; My loop has been executed 20 times.
By analogy, when my pageSize=100, my loop executes pageSize * pageNum times.
Is this correct?

**Comment by jxblum** [comment]
I will need to think more on this...

I was reading up on the Redis `SCAN` command ([doc](https://redis.io/commands/scan/)) more and came across a few more interesting reads. Search for:

* "Number of elements returned at every SCAN call"
* "The COUNT option" (logically following the first read, and it seems `COUNT` may only be a "hint")
* "Why SCAN may return all the items of an aggregate data type in a single call?"
* ...


Although, it seems like `SCAN` itself should do the right thing (because `SCAN` always uses a key space with hash tables, unlike `SSCAN`, `HSCAN` and `ZSCAN`)??

I need to write a test case to verify some things.

**Comment by wendao8469759** [comment]
Yes, according to the explanation in the official website doc。
Time complexity: O(1) for every call. O(N) for a complete iteration, including enough command calls for the cursor to return back to 0. N is the number of elements inside the collection.

**Comment by jxblum** [comment]
PART 1 (of 2):

After a more careful and thorough analysis, I have learned more about Redis's behavior as well as a limitation (questionably, and arguably, bug) in Spring Data Redis (SD Redis).

> NOTE: This will need the extended Spring Data team's review before we move forward on this ticket.

The limitation (bug?) in Spring Data Redis (SD Redis) is that it does not take the current "cursor" into account when subsequently running and repeating the Redis `SCAN` command.

If we trace the `SCAN` down from the `RedisTemplate`, starting [here](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/core/RedisTemplate.java#L643-L650), first you see that it uses a Redis connection object to run the Redis `SCAN` command ([here](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/core/RedisTemplate.java#L648)).

Ultimately, SD Redis is going to use a `KeyCommands` implementation (either from Jedis or Lettuce, thus the driver) to execute the Redis `SCAN` command using the Jedis or Lettuce connection, as seen [here](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/DefaultedRedisConnection.java#L135).

If we trace from Jedis, we see this [block of code](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/jedis/JedisKeyCommands.java#L132-L180) runs the `SCAN` using Jedis ([Javadoc](https://www.javadoc.io/static/redis.clients/jedis/4.3.2/redis/clients/jedis/Jedis.html#scan-byte:A-redis.clients.jedis.params.ScanParams-)), for instance, [here](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/jedis/JedisKeyCommands.java#L170) (typeless option).

We notice that the "cursorId" is the "cursor" argument to the `Jedis.scan(cursor:byte[], :ScanOptions)` call in this case. It is this "cursor" that will set the beginning for the next, or subsequent scan.

However, nothing calls the SD Redis essential `JedisKeyCommands.scan(..)` [method](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/jedis/JedisKeyCommands.java#L143) other than the [overloaded method](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/jedis/JedisKeyCommands.java#L134) above, which simply passes `0` for the "cursor", thereby always causing the scan to start at the beginning.

A similar outcome is revealed with tracing from Lettuce. Again, we see this [block of code](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceKeyCommands.java#L127-L169).

Ultimately, the limitation is induced by the SD Redis `ScanOptions` type ([Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/core/ScanOptions.html)), which does not given you an "option" for setting the current "cursor" (marker from a previous scan).

This is important since the SD Redis `ScanOptions` gets "[converted](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConverters.java#L772-L795)" to Lettuce `ScanArgs` before executing the `SCAN`, invoked from [here](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceKeyCommands.java#L157), and as seen [here](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceKeyCommands.java#L156).

As you can see in the [conversion](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConverters.java#L772-L795), only the "pattern", "count" (LIMIT) and (optionally) type are taken into consideration, which is arguably a limitation derived from Lettuce's `KeyScanArgs` type ([Javadoc](https://lettuce.io/core/release/api/io/lettuce/core/KeyScanArgs.html)) itself, which has no option to set the (current) "cursor" (position).

We actually see the same limitation with Jedis when [converting](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/jedis/JedisKeyCommands.java#L154) the SD Redis `ScanOptions` to Jedis `ScanParams` ([Javadoc](https://www.javadoc.io/static/redis.clients/jedis/4.3.2/redis/clients/jedis/params/ScanParams.html)).

Although, like Jedis, which allows the current "cursor" position to be invoked in the `Jedis.scan(..)` call, again as seen [here](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/jedis/JedisKeyCommands.java#L170), Lettuce only provides the means to pass the current "cursor" position into the Lettuce driver when the `SCAN` is invoked, by "customizing" (or reconstructing) the already constructed Lettuce `ScanCursor` object using the SD Redis `ScanOptions` object ([Javadoc](https://lettuce.io/core/release/api/io/lettuce/core/ScanCursor.html)), from [here](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceKeyCommands.java#L150), to [here](https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceKeyCommands.java#L157).

So, in summary on PART 1, is arguable and possible that Spring Data Redis could set the current position of the "cursor" for subsequent `SCANs`.

However, the question is, does it matter?

**Comment by jxblum** [comment]
PART 2:

Following on from PART 1 [above](https://github.com/spring-projects/spring-data-redis/issues/2584#issuecomment-1560060497), the question becomes, is setting the starting position of the cursor useful?

In your use case, specifically "paging" the results, I'd argue that using the Redis `SCAN` is not useful!

First of all, there is no guarantee that Redis will return the same order of the scanned results on subsequent scans.

Consider for example that you have a 6 element list of keys `[ A, B, C, D, E, F ]`, with a page size of `2`, then the cursor used to iterate the results of the `SCAN` is NOT guaranteed to return the same order of elements in subsequent scans (for the same pattern), even given the current cursor position.

For instance, let's say the full (first) `SCAN` yields `[ B, C, A, D, F, E ]`, then on the first `SCAN` (page size / LIMIT of 2) it would return [ B, C] for the first page, [ A, D ] for the second page, and [ F, E ] for the final page.

However, if your run only a SCAN to get the first page [ B, C ], then rerun the `SCAN` to get the second page with the previous returned cursor position and LIMIT of `2` to get the second page, it might return `[ B, A]` for the second page where `B` already appeared on the first page from the initial scan.

Additionally, as I mentioned previously the `LIMIT` ("count") of the `SCAN` command is just a suggestion, and actually might return 1 elements or 3 or more elements in a page size (`LIMIT`) of `2`.

Again, this was called out in the Redis `SCAN` command [docs](https://redis.io/commands/scan/), specifically in the "_Scan guarantees_", "_Number of elements returned at every SCAN call_" and "_The COUNT option_" sections.

EXAMPLE

I spent the better part of yesterday right tests in a new GitHub Repository ([redis-experiments](https://github.com/jxblum/redis-experiments)) I setup to experiment with, learn and test the behaviors of **Jedis**, **Lettuce** and **Spring Data Redis**, independently as well as together (for example: Spring Data Redis with either Jedis or Lettuce).

In the Jedis-based `RedisScanIntegrationTests` ([source](https://github.com/jxblum/redis-experiments/blob/spring-data-redis-issue-2584/jedis-experiments/src/test/java/io/vmware/spring/jedis/client/tests/RedisScanIntegrationTests.java)) I have 2 test case methods that test the 2 problems I cited above: a [test case for order](https://github.com/jxblum/redis-experiments/blob/spring-data-redis-issue-2584/jedis-experiments/src/test/java/io/vmware/spring/jedis/client/tests/RedisScanIntegrationTests.java#L178-L213) and a [test case for limit](https://github.com/jxblum/redis-experiments/blob/spring-data-redis-issue-2584/jedis-experiments/src/test/java/io/vmware/spring/jedis/client/tests/RedisScanIntegrationTests.java#L147-L176).

Of course, the test case to test order (`scanIsConsistent`) also tests limit.

Both tests can pass, however, most of the time, they fail.

While it is easy to accommodate the `LIMIT`, particularly if it exceeds your count (LIMIT), not so much when it is less, it is not possible to control ORDER.

Nothing in the [docs](https://redis.io/commands/scan/) for the Redis `SCAN` command reveal anything about the order in which the keys from the `SCAN` will be returned. Simply search for "order".

CONCLUSION

So, in conclusion, I think your use of Redis's `SCAN` command for paging is misplaced. Redis is simply not that smart.

**Comment by jxblum** [comment]
@mp911de - Can you please provide additional insights and thoughts on this matter?

I am reasonably confident my analysis is correct, but I recognize that my knowledge of Redis is not complete and may have missed something important.

Thank you!

**Comment by mp911de** [comment]
> When my pageSize=1 and pageNum=10; My loop has been executed 10 times.
> When my pageSize=2 and pageNum=10; My loop has been executed 20 times.
> By analogy, when my pageSize=100, my loop executes pageSize * pageNum times.
> Is this correct?

Yes, it is.

Spring Data Redis is built for providing a more convenient abstraction to cater common use cases. You're looking for another optimization where you can retain the cursor and reuse it the next time you want to resume scanning.

There are two problems with that:

1. The result size isn't uniform and Redis can return less or more data on each iteration
2. Therefore, you cannot easily take a portion of data and assume you hand in the cursor and resume from that exact position.

As John mentioned, we already expose `scan` methods accepting a cursor at `JedisKeyCommands`. Therefore, we can easily expose these to `RedisKeyCommands` as well enabling your use-case through the connection API.

**Comment by wendao8469759** [comment]
@jxblum @mp911de 
Thank you very much.

What you're saying is very correct. Redis scan itself has certain drawbacks, such as returning duplicate elements when calling scan twice
So I understand why spring data redis is doing this.

**Comment by jxblum** [comment]
I propose we change the SD Redis `ScanOptions` type ([Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/core/ScanOptions.html)) to expose an API to set the "cursor", default to `0` when not explicitly specified by the caller.

In this way, we do not need to expose more methods on the `RedisKeyCommands` interface (or higher up, in `RedisTemplate` / `RedisOperations`) that introduce yet another method with an overload signature.

Additionally, the `scan(..)`, whether using Jedis or Lettuce, actually wrap the `ScanOptions` argument to `scan(..)` in the returned `Cursor` ([Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/core/Cursor.html)), which already has a `getCursorId()` method ([Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/core/Cursor.html#getCursorId())).

In Spring Data Redis, the `Cursor` is the abstraction that represents the `ScanResult` object returned by the Jedis `scan(..)` method and the `KeyScanCursor` returned by the Lettuce scan.

Alternatively, we could, for completeness/convenience, offer a `scan(cursorId:String, :ScanOptions)` overloaded method on `RedisKeyCommands` interface, but I feel like this is unnecessary.

In my approach, you would simply:

```java

ScanOptions scanOptions = ScanOptions.builder()
    .match("..")
    .count(10)
    .build();


Cursor<?> cursor = redisTemplate.scan(scanOptions);

// process Cursor as necessary

ScanOptions nextScanOptions = ScanOptions.builder()
    .from(scanOptions)
    .withCursorId(cursor.getCursorId())
    .build();

Cursor<?> nextCursor = redisTemplate.scan(nextScanOptions);

// process Cursor as necessary
```

In this way the APIs of the pertinent data access methods (such as the `RedisTemplate`) do not need to change.

Using the `RedisKeyCommands` approach, you would:

```java
ScanOptions scanOptions = ScanOptions.builder()
    .match("...")
    .count(10)
    .build();

Cursor<?> cursor = redisConnection.keyCommands().scan(scanOptions);

// process Cursor as necessary

Cursor<?> nextCursor = redisConnection.keyCommands().scan(cursor.getCursorId(), scanOptions);

// process Cursor as necessary
```

I suppose either way is not bad, but... ???

**Comment by jxblum** [comment]
Following up to my posts earlier, I just wanted to say that the changes are still a WIP.

I started by implementing the overloaded `RedisKeyCommands` interface, `scan(..)` method accepting a cursor (ID) as described in the title of this issue despite my argument that changing the `ScanOptions` would be a better approach.

It turns out, at least in the context of SD Redis, that the `Cursor` abstraction does not do exactly as I would expect, and certainly does not honor the behavior of the scan result (cursor) returned by the Redis `SCAN` command as executed directly with either the Lettuce or Jedis, Redis client driver API, such as my [example](https://github.com/jxblum/redis-experiments/blob/spring-data-redis-issue-2584/jedis-experiments/src/test/java/io/vmware/spring/jedis/client/tests/RedisScanIntegrationTests.java#L147-L213) demonstrated, where it will return partial results given a cursor ID. However, inside SD Redis, it is all or nothing despite passing the cursor's current position. 

It's possible I did something wrong, but it's equally likely that the `Cursor` abstraction is not correct.

For instance, `scan(..)` should NOT be [called](https://github.com/spring-projects/spring-data-redis/blob/3.1.0/src/main/java/org/springframework/data/redis/core/ScanCursor.java#L176) inside the `ScanCursor`'s `hasNext()` method. This can lead to a StackOverflow if the state of the `Cursor` is out of sync.

The `Cursor`, much like a standard Java `Iterator`, should not advance the position of the cursor inside `hasNext()`; it should simply query whether results remain based on the current position of the iteration in the collection/stream of elements being iterated. Only `next()` is technically responsible for fetching the next batch/set of intermediate results.
This is precisely backward in SD Redis `ScanCursor` base class.

I ned to look at this more closely when I return next week to work. For time being, this will need to wait, unless @mp911de wants to have a look.

I put the changes on the topic/issue branch [issue/2584](https://github.com/spring-projects/spring-data-redis/tree/issue/2584) for the curious onlooker.

**Comment by mp911de** [comment]
The proposal to add the `cursorId` to `ScanOptions` makes a lot more sense. After all, `ScanOptions` denote the entry point to scanning. We need to keep in mind, that scanning in cluster mode will likely not work when providing a `cursorId` as cluster-scanning uses the node as part of its state.

**Comment by noojung** [comment]
Hi, I'd like to share a use case for this feature.

I'm using Spring Batch's `RedisItemReader`, which relies on `RedisTemplate.scan(ScanOptions)` and currently does not support restartability. Being able to resume scanning from a previously saved cursor would help avoid rescanning already traversed keys when restarting a failed batch job, especially for large keyspaces.

I understand that this would also require changes on the Spring Batch side and careful handling of Redis `SCAN` semantics.

Are there any plans to support specifying an initial cursor?

Thank you!
